Specialize optional tensor inputs to graphs in the JIT

t-vi opened the pull request:
This specializes optional tensor inputs to either a DimensionedTensorType or, when None is passed,
UndefinedTensor (aka AutogradZeroTensorType).
This works because we already have different specs and thus separate plans for the two cases.
It enhances the shape analysis - because now unwrapped optional tensors will have DimensionedTensorType with appropriate shape and required grad etc.
Also, when combined with "if-pruning" (which I understand #18259 works towards), we actually get much nicer concrete graphs, too.

t-vi committed:
Specialize optional tensor inputs to graphs in the JIT

This specializes optional tensor inputs to either a
DimensionedTensorType or, when None is passed,
UndefinedTensor (aka AutogradZeroTensorType).
This works because we already have different specs and thus
separate plans for the two cases.
It enhances the shape analysis - because now unwrapped optional
tensors will have DimensionedTensorType with appropriate shape
and required grad etc.
Also, when combined with "if-pruning" (which I understand

t-vi committed:
don't fall through switch!

eellison commented:
Looks great! 

I'm going to wait for @wanchaol to review since I have less of a knowledge of how UndefinedTensor works. I think maybe mustBeNone() api should check for UndefinedTensor case now but I'm completely sure

t-vi commented:
Well, I'll follow up if our patches don't achieve what I have in mind. :)

eellison commented:
Also i'm not completely sure on this either but Constant Propagation / Constant Pooling might need to be updated now to account for this case now too

t-vi commented:
Adding the input shapes happens right at the beginning, so I think it would be before the various passes.

eellison commented on test/test_jit.py:
```diff
@@ -4943,6 +4943,23 @@ def named_var_and(x, y):
+                return 0
+
+            fn(None)
+            g = fn.graph_for(None)
```
Add a check that the output of the function is 1 / 0 ?

t-vi replied:
Well, I only added a return value to not have it tell me that it needs to do nothing... :)

eellison replied:
ya more just as a sanity check that the optimizations i'm adding in this PR don't break this, although i'll wait for the landing of that PR with this one to be safe

t-vi replied:
What I would suggest is that we amend the check to see that the graph is only a constant return for each case. :)

eellison reviewed (approved):
I think this looks good. I will update my other PR so that it handles the Undefined Tensor case. Maybe wait to land until @wanchaol comments

wanchaol reviewed (approved):
lgtm, have a quick conversation with @eellison, this might need to deal with other types for shape analysis.

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -513,6 +513,12 @@ class ShapePropagator {
         return;
       }
+      case prim::unchecked_unwrap_optional: {
+	if (node->input()->type()->isSubtypeOf(TensorType::get())) {
```
The more general case here is to set the output to be the unwrapped type of the input if it's an optional input, and the input if it's not an optional

t-vi replied:
I think we need to be a bit more detailed than that about None. So I'd use the following:
- if it's an `IValue` and that `isNone`, leave the output alone,
- if it's an `OptionalType` return the `elementType`,
- otherwise return the Type.

t-vi replied:
Well, for unsafe, we could always pass on the type because it should only be inserted where we know it's not None, so I'll skip the first there.

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -529,6 +535,10 @@ class ShapePropagator {
         return;
       }
       case aten::_unwrap_optional: {
+	if (node->input()->type()->isSubtypeOf(TensorType::get())) {
```
Since _unwrap_optional will throw an exception if it's None, we can do the same thing as with prim::unchecked_unwrap and set the output to be the unwrapped type of the input if it's an optional input, and the input if it's not an optional

t-vi replied:
Same as above. Actually, the None handling is already in that case.

t-vi committed:
Refine conditions, fix test

Conditions as discussed in the review

With correct indenting, the test actually tests...

t-vi commented:
The CI failures seem unrelated, so if you're happy with it, I'd suggest to land this.

t-vi commented:
So Adam thinks it that it is preferrable to use None instead of UndefinedTensor. I'll follow up with an improvement.

apaszke commented on torch/csrc/jit/argument_spec.h:
```diff
@@ -153,7 +153,8 @@ struct ArgumentSpec {
   TypePtr fillType(TypePtr original, size_t& offset) const {
-    if (original->isSubtypeOf(TensorType::get())) {
+    if (original->isSubtypeOf(TensorType::get())
+	|| original->isSubtypeOf(OptionalType::ofTensor())) {
```
This is quite surprising. Do we seriously translate `None` to an undefined tensor when you call a script function? How can that even work?

apaszke replied:
Oh, is it because all args are zero-initialized, so you're abusing the fact that this arg is not a tensor, but it will report being undefined? Bleh

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -513,6 +513,16 @@ class ShapePropagator {
+	if(auto ot = node->input()->type()->cast<OptionalType>()) {
+	  node->output()->setType(ot->getElementType());
+	} else {
+	  node->output()->setType(node->input()->type());
```
Why would we have those calls on non-optional inputs?

apaszke replied:
Also, when is that op used?

eellison replied:
These calls would be inserted before input types have been specialized. If we have an optional tensor as an input to the graph and we specialize the graph to a non-optional tensor, than these calls will no longer be needed

zdevito replied:
When does an unchecked_unwrap_optional ever have a different type than ot->getElementType? This code looks like a no-op to me. Why is it ineeded?

t-vi replied:
So as this PR is dead and has been reverted, I tried to add a better high-level overview this in #18407 .
The gist is that I've been trying to convert Optional[Tensor] into DimensionedTensorType or NoneType once we know what has been input.
May I suggest we move the discussion to #18407?

apaszke commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -529,10 +539,17 @@ class ShapePropagator {
+	  node->output()->setType(ot->getElementType());
+	} else {
+	  node->output()->setType(node->input()->type());
+	}
```
Why did we add this? How is `_unwrap_optional` different from `unchecked_unwrap_optional`? This code looks really weird, because we're trying to see if we can infer statically that this thing gets a `None` as an argument (which it never should??), and otherwise we assume it doesn't?

t-vi replied:
Ok. I think we should revert this and probably also abandon the newer patch.

eellison replied:
_unwrap_optional is user code. it throws an error if the input is None and returns the input. unchecked_unwrap_optional has been inserted by the compiler when we can statically reason that the input will never be none (and so doesn't check if the input is None)

apaszke reviewed (commented):
I'm very confused by what has changed here
